File: tests/fixtures/project_ts/src/middleware/errorHandler.ts
/**
 * Global error handling middleware.
 */

import { Request, Response, NextFunction } from "express";
import { AppError, ValidationError } from "../utils/errors";

export interface ErrorResponse {
  error: string;
  statusCode: number;
  fields?: Record<string, string>;
  stack?: string;
}

export function errorHandler(
  err: Error,
  req: Request,
  res: Response,
  next: NextFunction
): void {
  const isDev = process.env.NODE_ENV === "development";

  if (err instanceof ValidationError) {
    const response: ErrorResponse = {
      error: err.message,
      statusCode: err.statusCode,
      fields: err.fields,
    };
    res.status(err.statusCode).json(response);
    return;
  }

  if (err instanceof AppError) {
    const response: ErrorResponse = {
      error: err.message,
      statusCode: err.statusCode,
      stack: isDev ? err.stack : undefined,
    };
    res.status(err.statusCode).json(response);
    return;
  }

  // Unhandled errors
  const response: ErrorResponse = {
    error: isDev ? err.message : "Internal server error",
    statusCode: 500,
    stack: isDev ? err.stack : undefined,
  };
  res.status(500).json(response);
}

export function notFoundHandler(req: Request, res: Response): void {
  res.status(404).json({
    error: `Route ${req.method} ${req.path} not found`,
    statusCode: 404,
  });
}

File: tests/fixtures/project_go/middleware/error_handler.go
package middleware

import (
	"errors"

	"example.com/project/utils"
)

// ErrorResponse is the structured error response returned to clients.
type ErrorResponse struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

// ErrorHandler translates application errors into structured ErrorResponse values.
// It recognizes AppError, NotFoundError, and ValidationError types.
func ErrorHandler(err error) *ErrorResponse {
	if err == nil {
		return nil
	}

	var appErr *utils.AppError
	if errors.As(err, &appErr) {
		return &ErrorResponse{
			Code:    appErr.Code,
			Message: appErr.Message,
		}
	}

	var notFoundErr *utils.NotFoundError
	if errors.As(err, &notFoundErr) {
		return &ErrorResponse{
			Code:    404,
			Message: notFoundErr.Error(),
		}
	}

	var validationErr *utils.ValidationError
	if errors.As(err, &validationErr) {
		return &ErrorResponse{
			Code:    400,
			Message: validationErr.Error(),
		}
	}

	// Unknown error — return generic 500.
	return &ErrorResponse{
		Code:    500,
		Message: "internal server error",
	}
}
